In [ ]:
# Import libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# Create a dataset
# Let´s build the values lists
value_n_list = ['Andrew', 'Phillip', 'John']
print(value_n_list)

value_w_list = [50,         78,      91.5] # [kg]
value_h_list = [1.76,      1.87,     1.69] # [m]

# Let´s use a dictionary ({key:value})
# with the format {"str":list}
data_dict = {"name":value_n_list,
             "weight":value_w_list,
             "height":value_h_list}


print(data_dict)

# Turn a dictionary into a dataframe
df = pd.DataFrame(data_dict)

print(df)
# Call a variable
df

['Andrew', 'Phillip', 'John']
{'name': ['Andrew', 'Phillip', 'John'], 'weight': [50, 78, 91.5], 'height': [1.76, 1.87, 1.69]}
      name  weight  height
0   Andrew    50.0    1.76
1  Phillip    78.0    1.87
2     John    91.5    1.69


,name,weight,height
0,Andrew,50.0,1.76
1,Phillip,78.0,1.87
2,John,91.5,1.69


Aquí tienes una explicación paso a paso del código de la celda anterior:

### 1. Importar librerías
```python
# Import libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
```
*   `import pandas as pd`: Importa la librería `pandas`, que es fundamental para trabajar con estructuras de datos como `DataFrames` (similares a tablas o hojas de cálculo). Se le da el alias `pd` para facilitar su uso.
*   `import matplotlib.pyplot as plt`: Importa la parte `pyplot` de la librería `matplotlib`, utilizada para crear visualizaciones estáticas e interactivas en Python. Se le da el alias `plt`.
*   `import numpy as np`: Importa la librería `numpy`, que proporciona soporte para arreglos y matrices grandes y operaciones matemáticas de alto nivel. Se le da el alias `np`.

### 2. Crear las listas de datos
```python
# Create a dataset
# Let´s build the values lists
value_n_list = ['Andrew', 'Phillip', 'John']
print(value_n_list)

value_w_list = [50,         78,      91.5] # [kg]
value_h_list = [1.76,      1.87,     1.69] # [m]
```
*   Se crean tres listas de Python:
    *   `value_n_list`: Contiene los nombres de las personas.
    *   `value_w_list`: Contiene los pesos de las personas en kilogramos.
    *   `value_h_list`: Contiene las alturas de las personas en metros.
*   `print(value_n_list)`: Muestra en la consola el contenido de la lista `value_n_list`.

### 3. Crear un diccionario de datos
```python
# Let´s use a dictionary ({key:value})
# with the format {"str":list}
data_dict = {"name":value_n_list,
             "weight":value_w_list,
             "height":value_h_list}

print(data_dict)
```
*   Se crea un diccionario llamado `data_dict`. Los diccionarios en Python almacenan datos en pares `clave:valor`.
*   En este caso, las claves son cadenas de texto (`"name"`, `"weight"`, `"height"`) y los valores son las listas que creamos anteriormente (`value_n_list`, `value_w_list`, `value_h_list`). Este formato es muy común para preparar datos antes de convertirlos a un `DataFrame` de pandas.
*   `print(data_dict)`: Muestra en la consola el contenido completo del diccionario.

### 4. Convertir el diccionario a un DataFrame de Pandas
```python
# Turn a dictionary into a dataframe
df = pd.DataFrame(data_dict)

print(df)
# Call a variable
df
```
*   `df = pd.DataFrame(data_dict)`: Esta es la parte crucial. Utiliza la función `DataFrame()` de pandas para convertir el diccionario `data_dict` en un objeto `DataFrame` y lo asigna a la variable `df`. Cada clave del diccionario se convierte en el nombre de una columna y su lista asociada se convierte en los datos de esa columna.
*   `print(df)`: Muestra en la consola el `DataFrame` completo.
*   `df`: En entornos como Colab o Jupyter Notebooks, simplemente escribir el nombre de una variable en la última línea de una celda hará que se muestre su representación amigable (en este caso, una tabla bien formateada del `DataFrame`).

In [ ]:
df["bmi"]=df["weight"]/(df["height"]**2)
df

,name,weight,height,bmi
0,Andrew,50.0,1.76,16.141529
1,Phillip,78.0,1.87,22.305471
2,John,91.5,1.69,32.036693


### 5. Calcular el Índice de Masa Corporal (BMI)
```python
df["bmi"]=df["weight"]/(df["height"]**2)
df
```
*   `df["bmi"] = df["weight"] / (df["height"]**2)`: Esta línea calcula el Índice de Masa Corporal (BMI) utilizando la fórmula peso / (altura^2). Crea una **nueva columna** en el DataFrame llamada `bmi` y asigna el resultado de este cálculo a cada fila.
*   `df`: Al colocar `df` al final de la celda, se muestra el DataFrame completo en la salida, incluyendo la nueva columna `bmi`.

In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   name    3 non-null      object 
 1   weight  3 non-null      float64
 2   height  3 non-null      float64
 3   bmi     3 non-null      float64
dtypes: float64(3), object(1)
memory usage: 228.0+ bytes


### 6. Obtener información concisa del DataFrame (`df.info()`)
```python
df.info()
```
*   `df.info()`: Este método imprime un resumen conciso de un DataFrame. Incluye información como:
    *   El tipo de objeto del DataFrame (`<class 'pandas.core.frame.DataFrame'>`).
    *   El número de entradas y el rango del índice.
    *   El número total de columnas.
    *   Una lista de las columnas, el número de valores no nulos (`Non-Null Count`) en cada una y su tipo de dato (`Dtype`).
    *   El número de columnas por tipo de dato.
    *   El uso de memoria del DataFrame.

Es muy útil para una inspección rápida de la estructura y la calidad de los datos, especialmente para identificar valores faltantes.

In [ ]:
df.shape

(3, 4)

### 7. Obtener la forma (dimensiones) del DataFrame (`df.shape`)
```python
df.shape
```
*   `df.shape`: Este atributo devuelve una tupla que representa las dimensiones del DataFrame. El primer elemento de la tupla es el número de filas y el segundo es el número de columnas. En este caso, `(3, 4)` significa que el DataFrame tiene 3 filas y 4 columnas.

In [ ]:
df.describe()

,weight,height,bmi
count,3.000000,3.000000,3.000000
mean,73.166667,1.773333,23.494564
std,21.167979,0.090738,8.014020
min,50.000000,1.690000,16.141529
25%,64.000000,1.725000,19.223500
50%,78.000000,1.760000,22.305471
75%,84.750000,1.815000,27.171082
max,91.500000,1.870000,32.036693


### 8. Obtener estadísticas descriptivas (`df.describe()`)
```python
df.describe()
```
*   `df.describe()`: Este método genera estadísticas descriptivas de las columnas numéricas en el DataFrame. Por defecto, solo se incluyen las columnas numéricas. La salida incluye:
    *   `count`: El número de valores no nulos.
    *   `mean`: La media de los valores.
    *   `std`: La desviación estándar.
    *   `min`: El valor mínimo.
    *   `25%`, `50%` (mediana), `75%`: Los percentiles (cuartiles).
    *   `max`: El valor máximo.

In [ ]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
weight,3.0,73.166667,21.167979,50.000000,64.0000,78.000000,84.750000,91.500000
height,3.0,1.773333,0.090738,1.690000,1.7250,1.760000,1.815000,1.870000
bmi,3.0,23.494564,8.014020,16.141529,19.2235,22.305471,27.171082,32.036693


### 9. Estadísticas descriptivas transpuesta (`df.describe().T`)
```python
df.describe().T
```
*   `df.describe().T`: Aquí, se llama primero a `df.describe()` para obtener las estadísticas descriptivas (como en el punto anterior) y luego se aplica `.T`. El método `.T` (transpose) intercambia las filas y las columnas del resultado. Esto hace que las estadísticas (count, mean, std, etc.) se conviertan en las columnas y los nombres de las columnas del DataFrame original se conviertan en las filas, lo que puede ser más fácil de leer en algunos casos.

In [ ]:
df.describe(include="O")

,name
count,3
unique,3
top,Andrew
freq,1


### 10. Estadísticas descriptivas para columnas de objetos (`df.describe(include="O")`)
```python
df.describe(include="O")
```
*   `df.describe(include="O")`: Al añadir `include="O"` (o `include='object'`), se le indica a `describe()` que solo muestre las estadísticas descriptivas para las columnas de tipo 'objeto' (generalmente cadenas de texto o datos categóricos). Para estas columnas, las estadísticas que se muestran son diferentes:
    *   `count`: El número de valores no nulos.
    *   `unique`: El número de valores únicos.
    *   `top`: El valor que aparece con más frecuencia.
    *   `freq`: La frecuencia (número de veces) del valor `top`.

In [ ]:
df.describe(include="all")

,name,weight,height,bmi
count,3,3.000000,3.000000,3.000000
unique,3,NaN,NaN,NaN
top,Andrew,NaN,NaN,NaN
freq,1,NaN,NaN,NaN
mean,NaN,73.166667,1.773333,23.494564
std,NaN,21.167979,0.090738,8.014020
min,NaN,50.000000,1.690000,16.141529
25%,NaN,64.000000,1.725000,19.223500
50%,NaN,78.000000,1.760000,22.305471
75%,NaN,84.750000,1.815000,27.171082


### 11. Estadísticas descriptivas para todas las columnas (`df.describe(include="all")`)
```python
df.describe(include="all")
```
*   `df.describe(include="all")`: Al utilizar `include="all"`, este método genera estadísticas descriptivas para **todas** las columnas del DataFrame, tanto numéricas como de tipo objeto. Para las columnas donde una estadística no aplica (por ejemplo, la media para una columna de texto o el 'top' para una columna numérica), se muestra `NaN` (Not a Number).

In [ ]:
s=

### ¿Qué es la Desviación Estándar (STD)?

La desviación estándar es una medida de la **dispersión** de un conjunto de datos. Nos indica cuánto se alejan, en promedio, los valores individuales de la media del conjunto de datos.

*   **Desviación estándar baja**: Los puntos de datos tienden a estar muy cerca de la media.
*   **Desviación estándar alta**: Los puntos de datos están más dispersos en un rango más amplio de valores, o más alejados de la media.

### Fórmula de la Desviación Estándar

La fórmula general para la desviación estándar de una **muestra** (que es la más comúnmente utilizada cuando se trabaja con datos) es:

$$\sigma = \sqrt{\frac{\sum_{i=1}^{N} (x_i - \mu)^2}{N-1}}$$

Donde:
*   $\sigma$ (sigma): Es la desviación estándar.
*   $\sum$ (sigma mayúscula): Indica la suma de todos los elementos.
*   $x_i$: Cada uno de los puntos de datos individuales.
*   $\mu$ (mu): Es la media (promedio) de todos los puntos de datos.
*   $N$: Es el número total de puntos de datos en la muestra.
*   $N-1$: Se usa en el denominador para dar una estimación insesgada de la desviación estándar de la población, especialmente cuando se trabaja con una muestra pequeña. Si tuviéramos la población completa, usaríamos $N$ en lugar de $N-1$.

### Explicación Paso a Paso para Calcular la Desviación Estándar

Vamos a usar un ejemplo sencillo para entenderlo mejor. Supongamos que tienes las siguientes edades de un grupo de personas: `[20, 22, 25, 28, 30]`.

1.  **Calcular la Media ($\mu$)**: Suma todos los valores y divídelos por el número de valores.
    
    $\mu = (20 + 22 + 25 + 28 + 30) / 5 = 125 / 5 = 25$
    
    La media es 25.
    
2.  **Calcular la Diferencia de cada punto de datos con la Media ($x_i - \mu$)**: Resta la media a cada valor individual.
    
    *   $20 - 25 = -5$
    *   $22 - 25 = -3$
    *   $25 - 25 = 0$
    *   $28 - 25 = 3$
    *   $30 - 25 = 5$
    
3.  **Elevar al Cuadrado cada Diferencia ($(x_i - \mu)^2$)**: Esto se hace para que los valores negativos no se cancelen con los positivos y para dar más peso a las desviaciones más grandes.
    
    *   $(-5)^2 = 25$
    *   $(-3)^2 = 9$
    *   $(0)^2 = 0$
    *   $(3)^2 = 9$
    *   $(5)^2 = 25$
    
4.  **Sumar todos los Cuadrados de las Diferencias ($\sum (x_i - \mu)^2$)**: Suma todos los resultados del paso anterior.
    
    $\sum (x_i - \mu)^2 = 25 + 9 + 0 + 9 + 25 = 68$
    
5.  **Dividir la Suma por $N-1$**: Donde $N$ es el número total de puntos de datos (en nuestro ejemplo, $N=5$).
    
    $68 / (5 - 1) = 68 / 4 = 17$
    
    Este valor (17) se conoce como la **varianza**.
    
6.  **Calcular la Raíz Cuadrada del Resultado**: Esto nos da la desviación estándar.
    
    $\sigma = \sqrt{17} \approx 4.12$
    
    Así, la desviación estándar de las edades `[20, 22, 25, 28, 30]` es aproximadamente **4.12**.

En tu DataFrame, cuando usas `df.describe()`, la columna `std` te muestra este valor calculado para cada una de las columnas numéricas (`weight`, `height`, `bmi`). Por ejemplo, para la columna `weight`, `std` es `21.167979`, lo que indica que los pesos de las personas en tu DataFrame varían bastante alrededor del peso medio.